src/ZPublisher/HTTPRequest.py

```diff
@@ -1053,10 +1053,14 @@ def get(self, key, default=None, returnTaints=0,
                 return URL
 
             if key == 'BODY' and self._file is not None:
-                p = self._file.tell()
+                try:
+                    fpos = self._file.tell()
+                except Exception:
+                    fpos = None
                 self._file.seek(0)
```

Comment:
You are right. I will look into this as well !

Reply:
@ale-rt I removed the `self._file.seek(0)` line as - in my understanding -`read()` takes the whole file either way.
`self._file.seek(fpos)` is in a condition (same as in other code portions) so doesn't cause issues.

This works with Gunicorn.

Reply:
Wouldn't it be better to cast the `Body` instance to a `BytesIO` one when it is assigned to `self._file`?
The other option is to patch `gunicorn` to add the `tell` and `seek` methods  in the `Body` class.

Reply:
> @ale-rt I removed the `self._file.seek(0)` line as - in my understanding -`read()` takes the whole file either way. `self._file.seek(fpos)` is in a condition (same as in other code portions) so doesn't cause issues.
> 
> This works with Gunicorn.

This is not the case with a `BytesIO` instance:
```
>>> from io import BytesIO
>>> a = BytesIO(b"ab")
>>> a.seek(1)
1
>>> a.read()
b'b'
>>> a.read()
b''
>>> a.seek(0)
0
>>> a.read()
b'ab'
```

Reply:
@ale-rt does this PR now look acceptable to you ?

If not, how would you see the casting of `Body` instance exactly ?

(by the way, I found [this discussion](https://github.com/benoitc/gunicorn/pull/2090) about `seek` and `tell` implementation in Gunicorn)

Reply:
BTW if we are sure that file is alfays a `BytesIO` instance, using `.getvalue` will simplify the code a lot:
```
>>> a.seek(2)
2
>>> a.getvalue()
b'ab'
>>> a.tell()
2
```

Reply:
> @ale-rt does this PR now look acceptable to you ?
> 
> If not, how would you see the casting of `Body` instance exactly ?
> 

It can be done [here](https://github.com/zopefoundation/Zope/blob/0704d5e2b474061d327dc0c2b13c5697627c9656/src/ZPublisher/HTTPRequest.py#L516), but that will definitely conflict with #1124.

> (by the way, I found [this discussion](https://github.com/benoitc/gunicorn/pull/2090) about `seek` and `tell` implementation in Gunicorn)

Yeah, it is a pity the work was not finalized.

Reply:
I know that we have other occurrences where something resembling `_file` was not something having a `.tell` method, but IMO this is something we can avoid here.

Anyway, I would like to hear @d-maurer opinion.
I do not feel competent enough to approve.